lib/utils/verify-signatures.js

```diff
@@ -0,0 +1,460 @@
+const cacache = require('cacache')
+const chalk = require('chalk')
+const crypto = require('crypto')
+const fetch = require('npm-registry-fetch')
+const jsonParse = require('json-parse-even-better-errors')
+const localeCompare = require('@isaacs/string-locale-compare')('en')
+const npa = require('npm-package-arg')
+const pacote = require('pacote')
+const path = require('path')
+const pickManifest = require('npm-pick-manifest')
+const table = require('text-table')
+
+const ansiTrim = require('../utils/ansi-trim.js')
+
+const validateSignature = async ({ message, signature, publicKey }) => {
+  const verifier = crypto.createVerify('SHA256')
+  verifier.write(message)
+  verifier.end()
+  return verifier.verify(publicKey, signature, 'base64')
+}
+
+class VerifySignatures {
+  constructor (tree, filterSet, npm, opts) {
+    this.tree = tree
+    this.filterSet = filterSet
+    this.npm = npm
+    this.opts = opts
+    this.edges = new Set()
+    this.keys = new Map()
+    this.invalid = new Set()
+    this.missing = new Set()
+    this.verified = 0
+    this.output = []
+    this.exitCode = 0
+  }
+
+  async run () {
+    // Find all deps in tree
+    const nodes = this.tree.inventory.values()
+    this.getEdges(nodes, 'edgesOut')
+
+    const edges = Array.from(this.edges)
+    const start = process.hrtime.bigint()
+    const registries = this.findAllRegistryUrls(edges, this.npm.flatOptions)
+
+    // Prefetch and cache public keys from the used registries
+    await Promise.all(registries.map(async (registry) => this.getKeys({ registry })))
+    await Promise.all(edges.map(async (edge) => await this.getVerifiedInfo(edge)))
```

Comment:
```suggestion
    await Promise.all(edges.map((edge) => this.getVerifiedInfo(edge)))
```
probably slightly faster to skip the unnecessary promise absorption steps